### PR TITLE
added html mail template

### PR DESCRIPTION
**Pull request by jdalsem**
- [x] tests
- [x] test page in dev tools
- [x] test mail (prevent dev tools to screen)
- [x] update template 
- [x] deal with recipient in footer
- [x] add email/address factory fromEntity that stores entity in address for future use returning addresses in Email->getTo
- [x] mail docs

### CODE REVIEW

File: actions/admin/site/settings.php [hunk 1]
```diff
@@ -99,8 +99,8 @@
 	}
 }
 
-$remove_branding = ('on' === get_input('remove_branding', false));
```
Comment (jeabakker) [inline actions/admin/site/settings.php:102]: what happened to the save?

File: docs/guides/email.rst [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+Below an overview of the feature of the email system.
+
+.. contents:: Contents
+   :local:
+   :depth: 1
+   
+HTML Mail
+=========
+
+As an admin you can configure your site to have all outgoing emails to be HTML emails or just plain text emails. This is enabled by default.
```
Comment (jeabakker) [inline docs/guides/email.rst:15]: 'HTML emails are enabled by default' is clearer

File: engine/classes/Elgg/Assets/ImageFetcherService.php [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+	 * @param string $image_url the image url to get
+	 *
+	 * @return false|array result contains
+	 * 	- data: the image data
+	 * 	- content-type: the content type of the image
+	 * 	- name: the name of the image
+	 */
+	public function getImage(string $image_url) {
+		if (empty($image_url)) {
+			return false;
```
Comment (jeabakker) [inline engine/classes/Elgg/Assets/ImageFetcherService.php:86]: why not throw an exception (invalid or missing argument)

File: engine/classes/Elgg/Assets/ImageFetcherService.php [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	/**
+	 * Load an image url from cache
+	 *
+	 * @param string $image_url the url to load
+	 *
+	 * @return array
+	 */
+	protected function loadFromCache(string $image_url): array {
+		
+		if (empty($image_url)) {
```
Comment (jeabakker) [inline engine/classes/Elgg/Assets/ImageFetcherService.php:145]: is this needed internally?

File: engine/classes/Elgg/Assets/ImageFetcherService.php [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+	 * Save image data in system cache for easy reuse
+	 *
+	 * @param string $image_url the image url
+	 * @param array  $data      the image data
+	 *
+	 * @return bool
+	 */
+	protected function saveToCache(string $image_url, array $data): bool {
+		
+		if (empty($data) || empty($image_url)) {
```
Comment (jeabakker) [inline engine/classes/Elgg/Assets/ImageFetcherService.php:164]: is this needed internally?

File: engine/classes/Elgg/Email/Address.php [hunk 6]
```diff
@@ -85,13 +90,37 @@ public function unsetName() {
 		$this->name = null;
 	}
 	
+	/**
+	 * Store the Entity related to this Address
+	 *
+	 * @param \ElggEntity $entity
+	 *
+	 * @return void'
```
Comment (jeabakker) [inline engine/classes/Elgg/Email/Address.php:98]: trailing `'`

File: engine/classes/Elgg/Email/Address.php [hunk 7]
```diff
@@ -85,13 +90,37 @@ public function unsetName() {
...
+	 * @since 4.0
+	 */
+	public function setEntity(\ElggEntity $entity): void {
+		$this->entity = $entity;
+	}
+	
+	/**
+	 * Returns the saved entity
+	 *
+	 * @return \ElggEntity|NULL
```
Comment (jeabakker) [inline engine/classes/Elgg/Email/Address.php:109]: lowercase `null` please

File: engine/classes/Elgg/EmailService.php [hunk 8]
```diff
@@ -166,36 +195,259 @@ public function transport(Email $email) {
...
 	 *
-	 * @param Email $email Email
+	 * @param MailMessage $message Current message
+	 * @param Email       $email   Email
 	 *
-	 * @return \Laminas\Mime\Message
+	 * @return \Laminas\Mail\Message
 	 */
-	protected function buildMessageBody(Email $email) {
+	protected function setMessageBody(MailMessage $message, Email $email): MailMessage {
```
Comment (jeabakker) [inline engine/classes/Elgg/EmailService.php:221]: this is not a correct function name, it doesn't set anything. It just prepares stuff

File: engine/classes/Elgg/Views/HtmlFormatter.php [hunk 9]
```diff
@@ -381,4 +384,64 @@ public function decode($string) {
 		);
 		return $string;
 	}
+	
+	/**
+	 * Adds inline style to html content
+	 *
+	 * @param string $html      html content
+	 * @param string $css       style text
+	 * @param string $body_only toggle to return the body contents instead of a full html
```
Comment (jeabakker) [inline engine/classes/Elgg/Views/HtmlFormatter.php:393]: `@param bool`

File: views/default/email/elements/body.php [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+			<table class="edge-wrapper">
+				<tr>
+					<td>
+						<table class="content-wrapper">
+							<tr>
+								<td>
+									<table class="header">
+										<tr>
+											<td>
+												<h1><{$body_title}</h1>
```
Comment (jeabakker) [inline views/default/email/elements/body.php:23]: additional `<`

File: views/default/email/email.css [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+.content {
+	background: white;
+}
+.content > tr > td {
+	padding: 20px 10px;
+}
+.footer td {
+	padding: 10px;
+}
```
Comment (jeabakker) [inline views/default/email/email.css:157]: please add new line

**Review by jeabakker — CHANGES_REQUESTED** [review]

File: engine/classes/Elgg/EmailService.php (hunk 8 above)
Comment (jdalsem) [reply engine/classes/Elgg/EmailService.php:221]: disagree... it sets the body on the message on line 259

**Review by jeabakker — APPROVED** [review]

**Pull request merged**